lib/jsdom/living/websockets/WebSocket-impl.js

```diff
@@ -0,0 +1,314 @@
+    let protocols = constructorArgs[1] !== undefined ? constructorArgs[1] : [];
+
+    const urlRecord = parseURL(url);
+    if (urlRecord === null) {
+      throw new DOMException(`The URL '${url}' is invalid.`, "SyntaxError");
+    }
+    if (urlRecord.scheme !== "ws" && urlRecord.scheme !== "wss") {
+      throw new DOMException(
+        `The URL's scheme must be either 'ws' or 'wss'. '${urlRecord.scheme}' is not allowed.`,
+        "SyntaxError"
+      );
+    }
+    if (urlRecord.fragment !== null) {
+      throw new DOMException(`The URL contains a fragment identifier ('${urlRecord.fragment}'). Fragment identifiers ` +
+                             "are not allowed in WebSocket URLs.", "SyntaxError");
+    }
+
+    if (typeof protocols === "string") {
+      protocols = [protocols];
+    }
+    const protocolSet = new Set();
+    for (const protocol of protocols) {
+      if (!verifySecWebSocketProtocol(protocol)) {
+        throw new DOMException(`The subprotocol '${protocol}' is invalid.`, "SyntaxError");
+      }
+      const lowered = protocol.toLowerCase();
+      if (protocolSet.has(lowered)) {
+        throw new DOMException(`The subprotocol '${protocol}' is duplicated.`, "SyntaxError");
+      }
+      protocolSet.add(lowered);
+    }
+
+    this._urlRecord = urlRecord;
+    this.url = serializeURL(urlRecord);
+    const nodeParsedURL = nodeURL.parse(this.url);
+    this.extensions = "";
+
+    this.binaryType = "blob";
+
+    this._ws = null;
+    // Used when this._ws has not been initialized yet.
+    this._readyState = CONNECTING;
+    this._requiredToFail = false;
+    this.bufferedAmount = 0;
+    this._sendQueue = [];
+
+    let openSocketsForWindow = openSockets.get(window._globalProxy);
+    if (openSocketsForWindow === undefined) {
+      openSocketsForWindow = new Set();
+      openSockets.set(window._globalProxy, openSocketsForWindow);
+    }
+    openSocketsForWindow.add(this);
+
+    openingQueues.set(this._ownerDocument, openingQueues.get(this._ownerDocument).then(() => new Promise(resolve => {
+      // close() called before _ws has been initialized.
+      if (this._requiredToFail) {
+        resolve();
+        this._readyState = CLOSED;
+        this._onConnectionClosed(1006, "");
+        return;
+      }
+
+      this._ws = new WebSocket(this.url, protocols, {
+        headers: {
+          "user-agent": window.navigator.userAgent,
+          cookie: this._ownerDocument._cookieJar.getCookieStringSync(nodeParsedURL, { http: true }),
+          origin: this._ownerDocument.origin
+        },
+        rejectUnauthorized: this._ownerDocument._strictSSL
+      });
+      this._ws.once("open", () => {
+        resolve();
+        this._onConnectionEstablished();
+      });
+      this._ws.on("message", this._onMessageReceived.bind(this));
+      this._ws.once("close", (...args) => {
+        resolve();
+        this._onConnectionClosed(...args);
+      });
+      this._ws.on("headers", headers => {
```

Comment:
This will be renamed to `'upgrade'` in version 4. Listeners will receive the `http.IncomingMessage` instance.